### ISSUE TITLE
WTS: How to include an additional button on a NavigationViewItem using Prism.Unity

### ISSUE THREAD

**Issue by mjmeans** [Microsoft/WindowsTemplateStudio#2535]
I'm wondering how to add an extra small button to a NavigationViewItem control in WTS. I'm also using Prism.Unity which was selected within the WTS wizard.

The behavior I'm looking for is like you see in the Microsoft Groove Music app. Look at the "Playlists" item. You see what is otherwise a normal NavigationViewItem with icon and text, but there is also a square [+] button that is right justified on the same line that operates a popup. Normally with code behind I would just add a NavigationViewItem.Content with a Grid and Button, but that doesn't work when using WTS because of the WTS implementation using resources.resw to specify the localized content of the NavigationViewItem.

Does anyone have a suggestion how this can be done in this case?

## System
* **VS Version: 2.3.18212.1
* **Windows Build: 10.0.17134.228

**Comment by mvegaca** [comment]
Hi @mjmeans we are having a look here, we will keep you updated, thank you :)

**Comment by mjmeans** [comment]
Thanks. I also noticed that the Groove Music app has a settings item that behaves similarly. It is right justified on a line with the user profile item. For the time being, I'm using this method until I find something better and more compatible with the WTS wizard. I'm wrapping the Navigation View in a Grid and placing this inside the grid:

```
        <VisualStateManager.VisualStateGroups>
            <VisualStateGroup>
                <VisualState>
                    <VisualState.StateTriggers>
                        <AdaptiveTrigger MinWindowWidth="{x:Bind navigationView.ExpandedModeThresholdWidth}" />
                    </VisualState.StateTriggers>
                    <VisualState.Setters>
                        <Setter Target="AddPlaylist.Margin" Value="0,-40,0,0"/>
                    </VisualState.Setters>
                </VisualState>
            </VisualStateGroup>
        </VisualStateManager.VisualStateGroups>

```

Then using the Add button like this `<NavigationViewItem x:Name="AddPlaylist" Icon="Add" Margin="0,0,0,0" HorizontalAlignment="Right"/>`. It will place the Add item to the right of the previous item on the line, and when below the expanded mode, it will put the Add button below the previous item.

I prefer a method that is based on a Style in Page.Resources or even a new property in NavigationViewItem or a subclass of NavigationViewItem. A Style is cleaner. But I haven't figured out how to do that in way that is agnostic to the base NavigationViewItem base style whenever .NET changes it.

**Comment by mvegaca** [comment]
Hi @mjmeans 

There is no problem in adding the NavigationView to a Grid. WTS wizard will still allow you to add Pages but you will have to add the NavigationViewItem manually.

We've created a PoC that shows how to adapt a WTS project to add the possibility to configure items with complex NavigationViewItem.Content (not only a simple string).

In this PoC you can see the Add Playlist button in the Playlist NavigationViewItem.

Now the NavigationView Header is now targeting to a new Header property. We've also updated the OnItemInvoked command to navigate if the NavigationViewItem is not a String.

[Issue2535PoC.zip](https://github.com/Microsoft/WindowsTemplateStudio/files/2345221/Issue2535PoC.zip).

Hope this can help you.

**Comment by mjmeans** [comment]
I downloaded and unblocked the Zip, then loaded in VS 2017 Community and when compiling I get the following error:
```
1>------ Rebuild All started: Project: Issue2535PoC, Configuration: Debug ARM ------
1>C:\Program Files (x86)\Windows Kits\10\bin\10.0.17134.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(280,5): Xaml Internal Error error WMC9999: Access is denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========

```

**Comment by mvegaca** [comment]
Try removing packages folder, Clean, Rebuild and Run the project

**Comment by mjmeans** [comment]
There is no packages folder within the solution or project. I tried closing VS and deleting both the obj and bin folders. Then reloading, clean and rebuild. The error returns on rebuild before trying run.

I then tried to add a new project in the same solution with the same layout; with the idea of moving over the source files from the PoC. Would not even create the project. Template 10 gave me an error: 

```
System.Exception: Error executing 'Microsoft.Templates.Core.PostActions.Catalog.Merge.MergePostAction'. Related template: Prism.Project.SplitView. ---> System.IO.InvalidDataException: Line namespace Issue2535PoC not found in file 'C:\Users\Mark\Documents\Visual Studio 2017\Projects\Issue2535PoC\App2\App.xaml.cs'. Related Template: 'Prism.Project.SplitView'.
   at Microsoft.Templates.Core.PostActions.Catalog.Merge.MergePostAction.ExecuteInternal()
   at Microsoft.Templates.Core.PostActions.PostAction.Execute()
   --- End of inner exception stack trace ---
   at Microsoft.Templates.Core.PostActions.PostAction.Execute()
   at Microsoft.Templates.UI.GenController.ExecutePostActions(GenInfo genInfo, TemplateCreationResult generationResult)
   at Microsoft.Templates.UI.GenController.<GenerateItemsAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Templates.UI.NewProjectGenController.<UnsafeGenerateProjectAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Templates.UI.NewProjectGenController.<GenerateProjectAsync>d__6.MoveNext()
```

I then tried to create a new WTS generated solution with the same layout to then port over the PoC files into. The generated solution built and ran; of course without the PoCs specific edits to the project. I then compared the generated solution file to the PoC solution file. No differences other than the expected GUIDs. I then compared the project files. I see that the PoC uses Microsoft.Toolkit.Uwp.UI, so I added it to the generated project. Presto. Build failure in the generated project. I then removed the Microsoft.Toolkit.Uwp.UI package from the generated project and it runs again.

So I've concluded that adding Microsoft.Toolkit.Uwp.UI 4.0.0 to a blank WTS project targeting Prism causes a build error. I don't know whether this is specific to my machine or not.

I tried removing the Microsoft.Toolkit.Uwp.UI package from the PoC. Now the PoC project builds and runs. I think this confirms that Microsoft.Toolkit.Uwp.UI 4.0.0 is causing the build failure.

Now with all that done... I have to say that the PoC solution doesn't behave the same as the Groove app. In non expanded mode, the [+] button disappears. It should appear as another icon in the hamburger menu even when not expanded. See the Windows 10 built-in Groove Music app. My workaround above is behaving correctly, though it may not be an optimal solution.

**Comment by sibille** [comment]
@mjmeans, I was able to reproduce your generation error adding a new project to a solution using "File->New Project-> Add to solution" while having selected the former project in the solution explorer. 
I opened a separate issue (#2541) and am working on it. While we're working on a solution, you can add a project to an existing solution from the solution context menu. 
Regarding Windows Community Toolkit 4.0.0 with Prism it's working fine on our machines. Could you try to refresh the nuget cache from the Nuget Package Manager Settings to see if this solves the problem?
![clearnugetcache](https://user-images.githubusercontent.com/24698018/45080241-c21d2980-b0f4-11e8-9368-d03f9a53a5c4.png)

**Comment by mjmeans** [comment]
I tried clearing the NuGet Cache(s) as you suggested. This solved the problem for the PoC.
